### ISSUE TITLE
Issue with calculate_area

### ISSUE THREAD

**Issue by sgupta92** [tobac-project/tobac#472]
I get the following error while using "tobac.calculate_area". At first, it looked like the issue was resolved with the latest release (1.5.4) but I still encounter it when using the function on files that were created using an older tobac release. 

After installing the latest release - the function runs when first processing the input file (dbz... .nc) using tobac but the area variable only includes "NaN" for all timesteps (see Track... .h5). When trying to re-run the function on the Track file, the same error pops up again. 

Input file and output files created after installing the latest release are uploaded here: https://drive.google.com/drive/u/0/folders/1ARx-GB59t9PXzRUJoD80jc07OIKp9OyS

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[82], line 1
----> 1 tobac.calculate_area(Features,Mask_dbz)

File ~/.local/lib/python3.11/site-packages/tobac/utils/decorators.py:279, in xarray_to_iris.<locals>.xarray_to_iris_i.<locals>.wrapper(*args, **kwargs)
    276             output = xarray.DataArray.from_iris(output)
    278 else:
--> 279     output = func(*args, **kwargs)
    280 # print(output)
    281 return output

File ~/.local/lib/python3.11/site-packages/tobac/analysis/spatial.py:440, in calculate_area(features, mask, method_area, vertical_coord)
    437 # Area needs to be a dataarray for get_statistics from mask, but otherwise dims/coords don't actually matter
    438 area = xr.DataArray(area, dims=("a", "b"))
--> 440 features = get_statistics_from_mask(
    441     features,
    442     mask,
    443     area,
    444     statistic={"area": np.sum},
    445     default=np.nan,
    446     collapse_dim=collapse_dim,
    447 )
    449 return features

File ~/.local/lib/python3.11/site-packages/tobac/utils/decorators.py:180, in iris_to_xarray.<locals>.iris_to_xarray_i.<locals>.wrapper(*args, **kwargs)
    177 kwargs_new = _conv_kwargs_iris_to_xarray(kwargs)
    178 # print(args)
    179 # print(kwargs)
--> 180 output = func(*args, **kwargs_new)
    181 if type(output) == tuple:
    182     output = tuple(
    183         [
    184             (
   (...)
    190         ]
    191     )

File ~/.local/lib/python3.11/site-packages/tobac/utils/bulk_statistics.py:326, in get_statistics_from_mask(features, segmentation_mask, statistic, index, default, id_column, collapse_dim, *fields)
    322         continue
    323     else:
    324         # make sure that features are not double-defined
    325         step_statistics.append(
--> 326             get_statistics(
    327                 features_t,
    328                 segmentation_mask_t,
    329                 *fields_t,
    330                 statistic=statistic,
    331                 default=default,
    332                 index=index,
    333                 id_column=id_column,
    334                 collapse_axis=collapse_axis,
    335             )
    336         )
    338 features = pd.concat(step_statistics)
    340 return features

File ~/.local/lib/python3.11/site-packages/tobac/utils/bulk_statistics.py:134, in get_statistics(features, labels, statistic, index, default, id_column, collapse_axis, *fields)
    131 index_in_features = np.isin(index, features[id_column])
    133 # set negative markers to 0 as they are unsegmented
--> 134 bins = np.cumsum(np.bincount(np.maximum(labels.ravel(), 0)))
    135 argsorted = np.argsort(labels.ravel())
    137 # Create lambdas to get (ravelled) label locations using argsorted and bins

File <__array_function__ internals>:200, in cumsum(*args, **kwargs)

File /opt/conda/lib/python3.11/site-packages/dask/array/core.py:1760, in Array.__array_function__(self, func, types, args, kwargs)
   1757 if has_keyword(da_func, "like"):
   1758     kwargs["like"] = self
-> 1760 return da_func(*args, **kwargs)

File /opt/conda/lib/python3.11/site-packages/dask/array/reductions.py:1617, in cumsum(x, axis, dtype, out, method)
   1604 @derived_from(np)
   1605 def cumsum(x, axis=None, dtype=None, out=None, method="sequential"):
   1606     """Dask added an additional keyword-only argument ``method``.
   1607 
   1608     method : {'sequential', 'blelloch'}, optional
   (...)
   1615           scheduler, and hardware.  More benchmarking is necessary.
   1616     """
-> 1617     return cumreduction(
   1618         np.cumsum,
   1619         _cumsum_merge,
   1620         0,
   1621         x,
   1622         axis,
   1623         dtype,
   1624         out=out,
   1625         method=method,
   1626         preop=np.sum,
   1627     )

File /opt/conda/lib/python3.11/site-packages/dask/array/reductions.py:1541, in cumreduction(func, binop, ident, x, axis, dtype, out, method, preop)
   1536     raise ValueError(
   1537         f'Invalid method for cumreduction.  Expected "sequential" or "blelloch".  Got: {method!r}'
   1538     )
   1540 if axis is None:
-> 1541     x = x.flatten().rechunk(chunks=x.npartitions)
   1542     axis = 0
   1543 if dtype is None:

File /opt/conda/lib/python3.11/site-packages/dask/array/core.py:2767, in Array.rechunk(self, chunks, threshold, block_size_limit, balance, method)
   2757 """Convert blocks in dask array x for new chunks.
   2758 
   2759 Refer to :func:`dask.array.rechunk` for full documentation.
   (...)
   2763 dask.array.rechunk : equivalent function
   2764 """
   2765 from dask.array.rechunk import rechunk  # avoid circular import
-> 2767 return rechunk(self, chunks, threshold, block_size_limit, balance, method)

File /opt/conda/lib/python3.11/site-packages/dask/array/rechunk.py:347, in rechunk(x, chunks, threshold, block_size_limit, balance, method)
    345 if isinstance(chunks, (tuple, list)):
    346     chunks = tuple(lc if lc is not None else rc for lc, rc in zip(chunks, x.chunks))
--> 347 chunks = normalize_chunks(
    348     chunks, x.shape, limit=block_size_limit, dtype=x.dtype, previous_chunks=x.chunks
    349 )
    351 # Now chunks are tuple of tuples
    352 ndim = x.ndim

File /opt/conda/lib/python3.11/site-packages/dask/array/core.py:3101, in normalize_chunks(chunks, shape, limit, dtype, previous_chunks)
   3098     chunks = tuple(c if c not in {None, -1} else s for c, s in zip(chunks, shape))
   3100 if chunks and shape is not None:
-> 3101     chunks = sum(
   3102         (
   3103             blockdims_from_blockshape((s,), (c,))
   3104             if not isinstance(c, (tuple, list))
   3105             else (c,)
   3106             for s, c in zip(shape, chunks)
   3107         ),
   3108         (),
   3109     )
   3110 for c in chunks:
   3111     if not c:

File /opt/conda/lib/python3.11/site-packages/dask/array/core.py:3103, in <genexpr>(.0)
   3098     chunks = tuple(c if c not in {None, -1} else s for c, s in zip(chunks, shape))
   3100 if chunks and shape is not None:
   3101     chunks = sum(
   3102         (
-> 3103             blockdims_from_blockshape((s,), (c,))
   3104             if not isinstance(c, (tuple, list))
   3105             else (c,)
   3106             for s, c in zip(shape, chunks)
   3107         ),
   3108         (),
   3109     )
   3110 for c in chunks:
   3111     if not c:

File /opt/conda/lib/python3.11/site-packages/dask/array/core.py:1260, in blockdims_from_blockshape(shape, chunks)
   1258     raise TypeError("Must supply shape= keyword argument")
   1259 if np.isnan(sum(shape)) or np.isnan(sum(chunks)):
-> 1260     raise ValueError(
   1261         "Array chunk sizes are unknown. shape: %s, chunks: %s%s"
   1262         % (shape, chunks, unknown_chunk_message)
   1263     )
   1264 if not all(map(is_integer, chunks)):
   1265     raise ValueError("chunks can only contain integers.")

ValueError: Array chunk sizes are unknown. shape: (nan,), chunks: (1,)

A possible solution: https://docs.dask.org/en/latest/array-chunks.html#unknown-chunks
Summary: to compute chunks sizes, use

   x.compute_chunk_sizes()  # for Dask Array `x`
   ddf.to_dask_array(lengths=True)  # for Dask DataFrame `ddf`

**Comment by sgupta92** [comment]
Hi @w-k-jones Thanks for the response! For some reason, the solutions you provided did not work (most likely because I didn't follow properly). But the issue you identified about passing a dask dataarray to the function did lead me to a similar workaround.

I was passing the segmentation mask as an iris cube (created during the initial tobac processing) so any conversion to dask dataarray must be occurring within bulk_statistics or calculate_area? I still got the error despite updating line 304 on my local bulk_statistics file (changing .data to .values) or passing the segmentation mask after calling dask.compute().

However, I did somehow find a solution:
> Mask_dbz.data
> tobac.calculate_area(Track,Mask_dbz)

Adding the first line loads the data into the memory and the calculate_area function runs fine! Thanks so much!
See more here: https://scitools-iris.readthedocs.io/en/latest/userguide/real_and_lazy_data.html

**Comment by w-k-jones** [comment]
Thanks for the update. I previously tested the data loaded using `xarray`, but can confirm the error is reproducible when loading mask_dbz as an iris cube. I expect we didn't run into this previously as by default `xarray.DataArray.from_iris` converts the data to a masked array (in the same way that `iris.Cube.data` does), but that can cause errors for integer data in older versions of xarray, so we switched to copying the data as is in `v1.5.4` which preserves the data as a dask array causing this error. I'll have a look into fixing this more generally

Regarding the workaround, [`scipy.ndimage.labeled_comprehension`](https://docs.scipy.org/doc/scipy-1.14.1/reference/generated/scipy.ndimage.labeled_comprehension.html) can be used to perform functions on labelled regions of an array much more efficiently, e.g. for this case:
```
Track["area"] = labeled_comprehension(
    Mask_dbz.data.ravel(), Mask_dbz.data.ravel(), Track.feature, lambda x: np.size(x)*0.25, np.float32, 0
)
```

**Comment by sgupta92** [comment]
Thanks for the better workaround, I'll try it! 

I should note I ran into the original error with some of the previous releases of tobac, as well. I just got back to working on this task recently and then posted the issue here.

**Comment by w-k-jones** [comment]
The original error should be fixed with the imminent `v1.5.5` release
